### PR TITLE
Logbandwidth

### PR DESCRIPTION
**Pull request by unconst**
-- adds a universal bittensor tblogger 
-- moves the mnist synapse to an ffnn synapse for convention
-- takes out the model toolbox where it is not needed yet
-- removes a faulty test from test_crypto (test an fail randomly by design)
--  change naming convention for ffnn synapse to the same as cifar
-- adds a config file for the ffnn synapse
-- allows models to be deep copied using a config

### CODE REVIEW

File: bittensor/synapse.py [hunk 1]
```diff
@@ -25,15 +25,25 @@ def synapse_key(self) -> str:
...
         SynapseClass = self.__class__
-        synapse_copy = SynapseClass()
+        
+        model_config = None
+        if config == None:
+            # If no passed config, try model config.
+            if self.config == None:
+                raise ValueError('Deep copy requires a passed model config object or a member model.config')
+            model_config = self.config
+        synapse_copy = SynapseClass(model_config)
```
Comment (shibshib) [inline bittensor/synapse.py]: this looks weird. `config` vs `self.config` can be confusing. Maybe rename config to `synapse_config`?

File: bittensor/synapses/dpn/config.py [hunk 2]
```diff
@@ -28,13 +30,13 @@ class DPNConfig:
...
 
-        >>> from bittensor.synapses.dpn.dpn_configuration import DPNConfig
+        >>> from bittensor.synapses.dpn.config import DPNConfig
 
         >>> # Initializing a DPN configuration
         >>> configuration = DPNConfig()
 
-        >>> # Accessing the model configuration
-        >>> configuration = model ( configuration )
+        >>> # Initializing a DNP Synapse
```
Comment (shibshib) [inline bittensor/synapses/dpn/config.py]: `DNP` -> `DPN`

File: bittensor/synapses/dpn/model.py [hunk 3]
```diff
@@ -17,22 +17,22 @@ class DPNSynapse(bittensor.Synapse):
...
-                    config: DPNConfig,
+                    config: DPNConfig = None,
                     dendrite: bittensor.Dendrite = None,
                     metagraph: bittensor.Metagraph = None
                 ):
         r""" Init a new DPN synapse module.
 
             Args:
-                config (:obj: `bittensor.dpn.dpn_configuration.DPNConfig`, `required`)
+                config (:obj: `bittensor.dpn.dpn_configuration.DPNConfig`, `required`, defaults to DPNConfig())
```
Comment (shibshib) [inline bittensor/synapses/dpn/model.py]: `bittensor.dpn.dpn_configuration.DPNConfig` should be `bittensor.dpn.config.DPNConfig`

File: bittensor/synapses/dpn/model.py [hunk 4]
```diff
@@ -17,22 +17,22 @@ class DPNSynapse(bittensor.Synapse):
     """
 
     def __init__(   self, 
-                    config: DPNConfig,
+                    config: DPNConfig = None,
```
Comment (shibshib) [inline bittensor/synapses/dpn/model.py:20]: should we allow this to be None? it'd break the model init

File: bittensor/utils/dispatcher.py [hunk 5]
```diff
@@ -113,6 +113,9 @@ def combine(self, expert_out, gates, multiply_by_gates=True):
         combined = zeros.index_add(0, batch_index, flat_stitched.float())
 
         # reshape as original
-        combined = combined.view(expert_out[0].shape)
+        try:
+          combined = combined.view(expert_out[0].shape)
+        except:
+          pass
```
Comment (shibshib) [inline bittensor/utils/dispatcher.py]: this is dangerous. Should at least give a logger error if we don't want to raise here.

File: bittensor/synapses/dpn/model.py (hunk 4 above)
Comment (unconst) [reply bittensor/synapses/dpn/model.py:20]: Shouldn't break because we have a check in the init

File: bittensor/utils/dispatcher.py (hunk 5 above)
Comment (unconst) [reply bittensor/utils/dispatcher.py]: This shouldn't be here.  Thanks

**Review by shibshib — APPROVED** [review]

**Pull request merged**